relay-metrics/src/aggregation.rs

```diff
@@ -1792,81 +1904,105 @@ impl Aggregator {
+    fn handle_merge_buckets(&mut self, msg: MergeBuckets) {
+        let MergeBuckets {
+            project_key,
+            buckets,
+        } = msg;
+        if let Err(err) = self.merge_all(project_key, buckets) {
+            relay_log::error!("failed to merge buckets: {}", LogError(&err));
+        }
+    }
 
-        // TODO: Consider a better approach than busy polling
-        ctx.run_interval(FLUSH_INTERVAL, |slf, _context| {
-            slf.try_flush();
-        });
+    fn handle_message(&mut self, msg: Aggregator) {
+        match msg {
+            Aggregator::AcceptsMetrics(_, sender) => self.handle_accepts_metrics(sender),
+            #[cfg(test)]
+            Aggregator::BucketCountInquiry(_, sender) => {
+                sender.send(self.buckets.len());
+            }
+            Aggregator::InsertMetrics(msg) => self.handle_insert_metrics(msg),
+            Aggregator::MergeBuckets(msg) => self.handle_merge_buckets(msg),
+        }
     }
 
-    fn stopped(&mut self, _ctx: &mut Self::Context) {
-        relay_log::info!("aggregator stopped");
+    fn handle_shutdown(&mut self, message: &Option<Shutdown>) {
+        if let Some(message) = message {
+            if message.timeout.is_some() {
+                self.state = AggregatorState::ShuttingDown;
+            }
+        }
     }
 }
 
-impl Default for Aggregator {
-    fn default() -> Self {
-        unimplemented!("register with the SystemRegistry instead")
+impl fmt::Debug for AggregatorService {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct(std::any::type_name::<Self>())
+            .field("config", &self.config)
+            .field("buckets", &self.buckets)
+            .field("receiver", &format_args!("Recipient<FlushBuckets>"))
+            .finish()
     }
 }
 
-impl Supervised for Aggregator {}
+impl Service for AggregatorService {
+    type Interface = Aggregator;
 
-impl SystemService for Aggregator {}
+    fn spawn_handler(mut self, mut rx: relay_system::Receiver<Self::Interface>) {
+        tokio::spawn(async move {
+            let mut ticker = time::interval(FLUSH_INTERVAL);
+            let mut shutdown = Controller::subscribe_v2().await;
+            relay_log::info!("aggregator started");
+            loop {
+                tokio::select! {
+                    biased;
 
-impl Handler<Shutdown> for Aggregator {
-    type Result = Result<(), ()>;
+                    _ = ticker.tick() => self.try_flush(),
+                    Some(message) = rx.recv() => self.handle_message(message),
+                    _ = shutdown.changed() => self.handle_shutdown(&shutdown.borrow_and_update()),
 
-    fn handle(&mut self, message: Shutdown, _context: &mut Self::Context) -> Self::Result {
-        if message.timeout.is_some() {
-            relay_log::trace!("Shutting down...");
-            self.state = AggregatorState::ShuttingDown;
-        }
-        Ok(())
+                    else => break,
```

Comment:
The shutdown logic of this loop eludes me, when does the loop ever break?  IIUC the select will only execute the else branch if all other branches are disabled.  To disable a branch it needs to have a precondition that evaluates to false, but none of the branches have a precondition.  So this loop never finishes?  I may be having this wrong (and if I have it wrong because it's not very obvious rather than I'm just being dumb maybe it's worth explaining it in a comment)

Reply:
So the new behaviour mirrors the old one for time being. This loop will exit only when the runtime stops - this how it was in the actix system. 

Once the rest of the actors migrate to the new tokio, we will refactor the shutdown process.

Reply:
Fair enough, mind adding a comment clarifying that this currently doesn't ever exit the loop?  Because I was wondering if I misunderstood this.

Reply:
yeah, will do 👍

Reply:
done in [9c2fb4f](https://github.com/getsentry/relay/pull/1508/commits/9c2fb4fd17801b0232ee48db13ea0d0ca8892df8)

Reply:
done in [9c2fb4f](https://github.com/getsentry/relay/pull/1508/commits/9c2fb4fd17801b0232ee48db13ea0d0ca8892df8)